File: execution-system/docker/python/execute.py
#!/usr/bin/env python3
import os
import sys
import resource
import signal
import traceback
from contextlib import contextmanager
import time

def get_memory_usage():
    """Get the memory usage of the current process in bytes"""
    return resource.getrusage(resource.RUSAGE_SELF).ru_maxrss * 1024

@contextmanager
def timeout(seconds):
    """Timeout context manager"""
    def signal_handler(signum, frame):
        raise TimeoutError("Code execution timed out")
    
    # Set the signal handler and alarm
    signal.signal(signal.SIGALRM, signal_handler)
    signal.alarm(seconds)
    try:
        yield
    finally:
        signal.alarm(0)

def main():
    try:
        # Get the code file path from arguments or default to /tmp/code.py
        code_file = sys.argv[1] if len(sys.argv) > 1 else '/tmp/code.py'
        
        if not os.path.exists(code_file):
            print(f"Error: Code file not found at {code_file}", file=sys.stderr)
            sys.exit(1)
        
        # Set resource limits
        # 128MB memory limit
        resource.setrlimit(resource.RLIMIT_AS, (128 * 1024 * 1024, 128 * 1024 * 1024))
        # 5 second CPU time limit
        resource.setrlimit(resource.RLIMIT_CPU, (5, 5))
        # 64MB file size limit
        resource.setrlimit(resource.RLIMIT_FSIZE, (64 * 1024 * 1024, 64 * 1024 * 1024))
        
        # Record start time and memory
        start_time = time.time()
        start_memory = get_memory_usage()
        
        # Execute the code with timeout
        with timeout(5):  # 5 second timeout
            # Create a new namespace for the code
            namespace = {}
            with open(code_file, 'r') as f:
                code = f.read()
                
            # Execute the code
            exec(code, namespace)
            
            # If there's a Solution class, try to run it
            if 'Solution' in namespace:
                solution = namespace['Solution']()
                # Look for a main method
                if hasattr(solution, 'main'):
                    solution.main()
        
        # Calculate resource usage
        end_time = time.time()
        end_memory = get_memory_usage()
        
        # Print execution stats as JSON
        execution_stats = {
            'runtime_ms': int((end_time - start_time) * 1000),
            'memory_bytes': end_memory - start_memory,
            'success': True
        }
        print(f"\n__EXECUTION_STATS__:{execution_stats}", file=sys.stderr)
        
    except TimeoutError:
        print("Error: Code execution timed out", file=sys.stderr)
        sys.exit(1)
    except MemoryError:
        print("Error: Memory limit exceeded", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"Error: {str(e)}", file=sys.stderr)
        print("Traceback:", file=sys.stderr)
        traceback.print_exc(file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    main() 

File: execution-system/docker/javascript/execute.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const v8 = require('v8');
const { performance } = require('perf_hooks');

// Set memory limit to 128MB
const MEMORY_LIMIT = 128 * 1024 * 1024;
v8.setFlagsFromString('--max_old_space_size=128');

// Default code file path
const DEFAULT_CODE_FILE = '/tmp/code.js';

// Get initial memory usage
const getMemoryUsage = () => {
    const used = process.memoryUsage();
    return used.heapUsed + used.external;
};

async function executeCode(codePath) {
    try {
        // Use provided path or default
        const codeFile = codePath || DEFAULT_CODE_FILE;

        // Check if file exists
        if (!fs.existsSync(codeFile)) {
            console.error(`Error: Code file not found at ${codeFile}`);
            process.exit(1);
        }

        const startTime = performance.now();
        const startMemory = getMemoryUsage();

        // Read the code file
        const code = fs.readFileSync(codeFile, 'utf8');

        // Create a new context for execution
        const context = {
            console: {
                log: (...args) => {
                    console.log(...args);
                },
                error: (...args) => {
                    console.error(...args);
                }
            },
            process,
            require,
            setTimeout,
            setInterval,
            clearTimeout,
            clearInterval,
            Buffer,
            URL,
            Error,
            Date,
            Math,
            JSON,
            // Add any other required globals
        };

        // Set execution timeout
        const timeoutPromise = new Promise((_, reject) => {
                setTimeout(() => {
                reject(new Error('Execution timeout'));
            }, 5000); // 5 seconds timeout
        });

        // Execute the code
        const executionPromise = new Promise(async (resolve) => {
            try {
                // Create function from code
                const fn = new Function('context', `
                    with (context) {
                        ${code}
                        // If Solution class exists, try to run it
                        if (typeof Solution !== 'undefined') {
                            const solution = new Solution();
                            if (typeof solution.main === 'function') {
                                solution.main();
                            }
                        }
                    }
                `);

                // Execute in context
                fn(context);
                resolve();
            } catch (error) {
                throw error;
            }
        });

        // Race between execution and timeout
        await Promise.race([executionPromise, timeoutPromise]);

        // Calculate resource usage
        const endTime = performance.now();
        const endMemory = getMemoryUsage();

        // Print execution stats
        const stats = {
            runtime_ms: Math.round(endTime - startTime),
            memory_bytes: endMemory - startMemory,
            success: true
        };
        console.error(`\n__EXECUTION_STATS__:${JSON.stringify(stats)}`);

    } catch (error) {
        if (error.message === 'Execution timeout') {
            console.error('Error: Code execution timed out');
        } else {
            console.error(`Error: ${error.message}`);
            console.error('Stacktrace:');
            console.error(error.stack);
        }
        process.exit(1);
    }
}

// Get code file path from arguments or use default
const codePath = process.argv[2] || DEFAULT_CODE_FILE;
executeCode(codePath).catch(error => {
    console.error(`Fatal error: ${error.message}`);
    process.exit(1);
}); 